### PR TITLE
enable segmentation on test data and generate image masks and submiss…

### PR DESCRIPTION
**Pull request by YiwenShaoStephen**
changes including:
(1) use kaggle instead of wget to download data. Previously I downloaded the data and uploaded it to amazon s3 for wget downloading. But it seems not obey the license since only kaggle users can download this data. In addition, we will need "kaggle" api for submitting results. So I changed it to kaggle here.
kaggle can be easily installed by pip install kaggle. But we need to put our account info in ~/.kaggle as [https://github.com/Kaggle/kaggle-api](url) instructs. I haven't updated dependency-related stuff. I think we can finish it later.
(2) now we can run segmentation on all test data (stage1_test) and get a run-length encoding file as the result. The segmentation results for each image is also stored on the disk thus we can better get an intuition on the quality of the segmentation.

TODO:
There are actually three data splits in total for this competition: train, stage1_test, stage2_test. Previously we only download train (further split into train and val) and stage1_test. Both of these two parts are provided with a ground truth in a csv file format (same way as I store). Stage2_test is the final test data without ground truth accessible yet. We need to write our segmentation result in csv file format and submit it to their official competition website [https://www.kaggle.com/c/data-science-bowl-2018/data](url) or via `kaggle competitions submit -c data-science-bowl-2018 -f submission.csv -m "Message"`. Stage2_test (3019) is also much larger compared to stage1_test (65).  So for now I think we'd better use stage1_test to tune the segmenter or network stuff. 
In conclusion, if you want to see our final results and compare it with other submission on leadboard,  just change `--test-data data/stage1_test` to `--test-data data/stage2_test_final` in run.sh, but it will take a quite long time (haven't done parallelization yet). Otherwise, if you want to tune the setup and do a quick test, you can use the default stage1_test. But for now, the scoring code is not finished yet, you can only see the encoded results and the visualized masks.

### CODE REVIEW

File: egs/dsb2018/v1/local/process_data.py [hunk 1]
```diff
@@ -105,42 +111,37 @@ def DataProcess(input_path, output_dir, channels, mode='train', train_prop=0.9):
 
     else:
         # Get test images
-        print('Getting test images ... ')
-        test_ids = next(os.walk(input_path))[1]
-        test = []
+        test_saver = DataSaver(os.path.join(
```
Comment (danpovey) [inline egs/dsb2018/v1/local/process_data.py]: It's not an issue with this code, but I just noticed that it looks like we are still not doing the compression?
Perhaps we could just build that into the DataSaver and loader classes rather than do it as I mentioned before with the compressed_image_with_mask object.

File: scripts/waldo/data_transformation.py [hunk 2]
```diff
@@ -90,7 +91,7 @@ def scale_down_image_with_objects(image_with_objects, config, max_size):
     return resized_image_with_objects
 
 
-def make_square_image_with_padding(im_arr, config):
+def make_square_image_with_padding(im_arr, num_colors):
```
Comment (danpovey) [inline scripts/waldo/data_transformation.py]: For this function (and this is a pre-existing issue), please clarify in the documentation whether it pads equally on both sides or just on the (top/left) or (bottom/right).

File: egs/dsb2018/v1/local/tuning/run_unet_1a.sh [hunk 3]
```diff
@@ -11,13 +11,13 @@ epochs=10
 start_filters=64
 batch_size=16
 lr=0.01
+dir=
```
Comment (danpovey) [inline egs/dsb2018/v1/local/tuning/run_unet_1a.sh]: I think this variable should have a default, e.g. exp/unet_1a
These 'run'-type scripts should always do something reasonable when run without options.

File: egs/dsb2018/v1/local/prepare_data.sh [hunk 4]
```diff
@@ -9,21 +9,17 @@
...
+test2_dir=$dl_dir/stage2_test_final
 
 mkdir -p $dl_dir
 if [ -d $train_dir ]; then
   echo Not downloading DSB2018 training data as it is already there.
 else
   if [ ! -f $dl_dir/stage1_train.zip ]; then
-    echo Downloading DSB2018 training data...
-    wget -P $dl_dir $train_url || exit 1;
+    kaggle competitions download -c data-science-bowl-2018 -f stage1_train.zip -p $dl_dir
```
Comment (danpovey) [inline egs/dsb2018/v1/local/prepare_data.sh]: I assume you should add 'kaggle' to requirements.txt in egs/dsb2018/v1?

File: egs/dsb2018/v1/local/segment.py [hunk 5]
```diff
@@ -3,23 +3,24 @@
...
 from unet_config import UnetConfig
 
 
 parser = argparse.ArgumentParser(description='Pytorch DSB2018 setup')
-parser.add_argument('test_data', default='./data', type=str,
-                    help='Path to processed validation data')
-parser.add_argument('dir', type=str,
+parser.add_argument('--test-data', default='data/stage1_test', type=str,
+                    help='Path to test images to be segmented')
+parser.add_argument('--dir', type=str,
```
Comment (hhadian) [inline egs/dsb2018/v1/local/segment.py]: Is there a reason you are making `dir` and `test_data` optional? They are required.

File: egs/dsb2018/v1/local/tuning/run_unet_1a.sh [hunk 6]
```diff
@@ -28,7 +28,7 @@ if [ $stage -le 1 ]; then
   num_classes $num_classes
   num_colors $num_colors
   padding $padding
-  offsets 1 0  0 1  -2 -1  1 -2
+  offsets 1 0  0 1  -2 -1  1 -2
```
Comment (hhadian) [inline egs/dsb2018/v1/local/tuning/run_unet_1a.sh]: I guess you should configure your text editor to remove trailing spaces. This will avoid unwanted space-related changes from being commited.

File: egs/dsb2018/v1/local/segment.py (hunk 5 above)
Comment (danpovey) [reply egs/dsb2018/v1/local/segment.py]: Yiwen, just because others jump off a cliff, doesn't mean we have to as well.  Please either make them positional arguments (like 'dir') or make them use option syntax (like '--dir') but with required=True.

Comment (YiwenShaoStephen) [reply egs/dsb2018/v1/local/segment.py]: Oh, yes. I forget there is a required option. I will add that.

File: egs/dsb2018/v1/local/prepare_data.sh (hunk 4 above)
Comment (YiwenShaoStephen) [reply egs/dsb2018/v1/local/prepare_data.sh]: Sure, I will do it if you're ok with using kaggle.

File: egs/dsb2018/v1/local/tuning/run_unet_1a.sh (hunk 3 above)
Comment (YiwenShaoStephen) [reply egs/dsb2018/v1/local/tuning/run_unet_1a.sh]: No problem

File: egs/dsb2018/v1/local/process_data.py (hunk 1 above)
Comment (YiwenShaoStephen) [reply egs/dsb2018/v1/local/process_data.py]: OK, will have a try.

**Comment by YiwenShaoStephen** [comment]
@aarora8 @ChunChiehChang Now you can use WaldoDataset to get ground truth mask with mask=True option when you initialize it.

**Comment by danpovey** [comment]
merging this, if there outstanding issues you can merge them later.
